Migrate TestShowMoreCard.js

bobsilverberg opened the pull request:
Fixes #11252

bobsilverberg commented:
Note that the coverage failure is expected and is discussed in a comment in the test suite, and will be addressed via #11409.

eviljeff commented on src/amo/components/ShowMoreCard/index.js:
```diff
@@ -23,20 +23,48 @@ type UIStateType = {|
+  uiState: UIStateType,
+|};
+
+export const truncateToMaxHeight: (TruncateToMaxHeightProps) => void = ({
+  contents,
+  maxHeight,
+  setUIState,
+  uiState,
+}: TruncateToMaxHeightProps) => {
```
Why do we have to return all the same parameters that were passed in?  (and so necessitating a weird `TruncateToMaxHeightProps` type.)

bobsilverberg replied:
We don't. It's just that the Flow definitions are a bit hard to follow. I realized we don't need to define the type of the function itself, which removes some of the code above and hopefully makes it clearer. It also means we don't need the extra  `TruncateToMaxHeightProps` type.

eviljeff reviewed (commented):
Can you add some explanation of what's going on with the rewrite of `truncateToMaxHeight`?  The test changes themselves are pretty straightforward in comparison.

bobsilverberg commented on src/amo/components/ShowMoreCard/index.js:
```diff
@@ -23,20 +23,46 @@ type UIStateType = {|
   readMoreExpanded: boolean,
 |};
 
+export const truncateToMaxHeight = ({
```
This function was moved outside of the component's definition so that it can be the default value for `_truncateToMaxHeight` inside the component.

eviljeff replied:
mmm, but just moving it outside the component changes:
```
truncateToMaxHeight: (contents: HTMLElement | null) => void = (
    contents: HTMLElement | null,
  ) =>
```
to
```
export const truncateToMaxHeight = ({
  contents,
  maxHeight,
  setUIState,
  uiState,
}: {|
  contents: HTMLElement | null,
  maxHeight: number,
  setUIState: ($Shape<UIStateType>) => void,
  uiState: UIStateType,
|}) =>
```
but he functionality hasn't changed?

bobsilverberg replied:
Correct, the functionality has not changed. We now need to accept arguments for `maxHeight`, `setUIState` and `uiState` because previously those were being read from `this.props`, but `this.props` doesn't exist when we move the function outside of the component.

bobsilverberg commented on src/amo/components/ShowMoreCard/index.js:
```diff
@@ -23,20 +23,46 @@ type UIStateType = {|
 |};
 
 type InternalProps = {|
+  _setUIState?: ($Shape<UIStateType>) => void,
```
We also need to be able to mock `setUIState` from a test, which is why this has become a new prop.

eviljeff replied:
so `_setUIState` only exists to be able to override `setUIState` in tests?  it's pretty weird to me that's a necessity in our testing setup but okay.

bobsilverberg commented on src/amo/components/ShowMoreCard/index.js:
```diff
@@ -53,19 +79,38 @@ export class ShowMoreCardBase extends React.Component<InternalProps> {
+      _truncateToMaxHeight({
+        contents: this.contents,
+        maxHeight,
+        setUIState: _setUIState || setUIState,
```
We cannot use `defaultProps` to set a default value for `setUIState` because the default is a prop that is injected via the `withUIState` HOC, and is therefore not available when `defaultProps` executes. This is why we need this conditional here and also in `componentDidUpdate`.

bobsilverberg commented:
> Can you add some explanation of what's going on with the rewrite of `truncateToMaxHeight`? The test changes themselves are pretty straightforward in comparison.

I added a few comments that hopefully explain the changes in `ShowMoreCard`. I also changed the Flow definitions for `truncateToMaxHeight` so that they are hopefully easier to follow.

eviljeff reviewed (approved):
still seems like a lot of magic syntax changes to work around Jest or whatever but if that's the way to do it then 🤷
